## Optimizacion combinatoria

In [1]:
#! pip install cvxpy
import cvxpy as cp
import numpy as np

### Ejemplo 0


In [25]:
import cvxpy as cp
import numpy as np

x = cp.Variable(shape=(2,1), name="x")

objective = cp.Maximize(4*x[0] + 5*x[1])

# Relajación lineal continua
constraints = [
    4*x[0] + 6*x[1] <= 24,
    2*x[0] + x[1] <= 6,
    x[0] - x[1] <= 1,
    x[0] <= 2,
    x[0] <= 1, # Ramificar x_1
    x[1] >= 4, # Ramificar x_2
    x >= 0
]

In [26]:
problem = cp.Problem(objective, constraints)
solution = problem.solve()

print("Valor máximo:", np.round(solution, 4))
print("Solución:\n", np.round(x.value, 4))

Valor máximo: 20.0
Solución:
 [[0.]
 [4.]]


Tenemos un elemento que no es entero: $x_1$  
$\Rightarrow$ Ocupamos ramificar, añadiendo dos modelos distintos con las restricciones $x_1\leq 1$ y $x_1\geq 2$.  

* Con $x_1 \geq 2$ resulta en la solución $x^T = (2,2)$ y $z^* = 18$

* Con $x_1 \leq 1$ resulta en la solución $x^T = (1,3.333)$ y $z^* = 20.66$
    - Se obtiene una solución no entera para $x_2 \Rightarrow$ volver a ramificar con $x_2<=3$ y $x_2>=4$ 
        1. $x^T = (1,3)$ y $z^* = 19$
        2. $x^T = (0,4)$ y $z^* = 20$


Por lo tanto, nos quedamos con la solución  $x^T = (0,4)$ y $z^* = 20$


In [28]:
# Forzando enteros
x = cp.Variable(shape=(2,1), name="x", integer=True)

objective = cp.Maximize(4*x[0] + 5*x[1])

# Relajación lineal continua
constraints = [
    4*x[0] + 6*x[1] <= 24,
    2*x[0] + x[1] <= 6,
    x[0] - x[1] <= 1,
    x[0] <= 2,
    x >= 0
]
problem = cp.Problem(objective, constraints)
solution = problem.solve()

print("Valor máximo:", np.round(solution, 4))
print("Solución:\n", np.round(x.value, 4))

Valor máximo: 20.0
Solución:
 [[-0.]
 [ 4.]]


### Ejemplo

In [42]:
x = cp.Variable(shape=(2,1), name="x", integer=True)

objective = cp.Maximize(8*x[0] + 5*x[1])

# Relajación lineal continua
constraints = [
    1*x[0] + 1*x[1] <= 6,
    9*x[0] + 5*x[1] <= 45,
    x >= 0
]
problem = cp.Problem(objective, constraints)
solution = problem.solve()

print("Valor máximo:", np.round(solution, 4))
print("Solución:\n", np.round(x.value, 4))

Valor máximo: 40.0
Solución:
 [[ 5.]
 [-0.]]


### Otro ejemplote

In [2]:
# Relajación lineal continua
x = cp.Variable(shape=(2,1), name="x")

objective = cp.Maximize(2*x[0] + 3*x[1])

constraints = [
    -3*x[0] + x[1] <= 1,
    4*x[0] + 2*x[1] <= 15,
    4*x[0] - x[1] <= 10,
    -x[0] + 2*x[1] <= 5,
    x >= 0
]

In [ ]:
problem = cp.Problem(objective, constraints)
solution = problem.solve()
print("Valor máximo:", np.round(solution, 4))
print("Solución:\n", np.round(x.value, 4))

Valor máximo: 14.5
Solución:
 [[2. ]
 [3.5]]


**Relajación Lineal:** $z^* = 14.5$, $x^* = (2,3.5)^T$  
Entonces, ahora probamos ramificando en $x_2$: $x_2 \leq 3$ y $x_2 \geq 4$

In [7]:
# x_2 <= 3
print("x_2 <= 3")
constraints = [
    -3*x[0] + x[1] <= 1,
    4*x[0] + 2*x[1] <= 15,
    4*x[0] - x[1] <= 10,
    -x[0] + 2*x[1] <= 5,
    x[1] <= 3, # Ramificar x_2
    x >= 0
]
problem = cp.Problem(objective, constraints)
solution = problem.solve()
print("Valor máximo:", np.round(solution, 4))
print("Solución:\n", np.round(x.value, 4))

# Ahora x_2 <= 3 y x_1 <= 2
print("x_2 <= 3 y x_1 <= 2")
constraints = [
    -3*x[0] + x[1] <= 1,
    4*x[0] + 2*x[1] <= 15,
    4*x[0] - x[1] <= 10,
    -x[0] + 2*x[1] <= 5,
    x[1] <= 3, # Ramificar x_2
    x[0] <= 2, # Ramificar x_1
    x >= 0
]
problem = cp.Problem(objective, constraints)
solution = problem.solve()
print("Valor máximo:", np.round(solution, 4))
print("Solución:\n", np.round(x.value, 4))

# Ahora x_2 <= 3 y x_1 >= 3
# NO jala


x_2 <= 3
Valor máximo: 13.5
Solución:
 [[2.25]
 [3.  ]]
x_2 <= 3 y x_1 <= 2
Valor máximo: 13.0
Solución:
 [[2.]
 [3.]]


In [9]:
# x_2 >= 4
# NO JALA

$\therefore z^* = 13$ y $x^* = (2,3)^T$

### Tarea 2

In [2]:

x = cp.Variable(shape=(4,1), name="x")

objective = cp.Maximize(10*x[0] + 15*x[1] + 4*x[2] + 2*x[3])

# Relajación lineal continua
constraints = [
    10*x[0] + 20*x[1] + 2*x[2] + 3*x[3] <= 4000,
    5*x[0] + 5*x[1] + 5*x[2] + 4*x[3] <= 1500,
    4*x[0] + 2*x[1] + 6*x[2] + 6*x[3] <= 800,
    x >= 0
]
problem = cp.Problem(objective, constraints)
solution = problem.solve()

print("Valor máximo:", np.round(solution, 4))
print("Solución:\n", np.round(x.value, 4))

Valor máximo: 3333.3333
Solución:
 [[133.3333]
 [133.3333]
 [  0.    ]
 [  0.    ]]


In [3]:
constraints += [x[0] + 2*x[1] + 3*x[2] + 4*x[3] >= 500]
problem = cp.Problem(objective, constraints)
solution = problem.solve()

print("Valor máximo:", solution)
print("Solución:\n", x.value)

Valor máximo: 3249.9999992452117
Solución:
 [[6.42857142e+01]
 [1.64285714e+02]
 [3.57142857e+01]
 [4.98201311e-08]]


In [10]:
import decimal, fractions
decimal_value = 64.2857136
decimal_object = decimal.Decimal(decimal_value)
fraction_object = fractions.Fraction.from_decimal(decimal_object)

numerator, denominator = fraction_object.as_integer_ratio()

print("%s / %s = %s" % (numerator, denominator, decimal_value))

4523704934596975 / 70368744177664 = 64.2857136


In [15]:
print(500/63)
print(11000/63)
print(1000/21)

7.936507936507937
174.6031746031746
47.61904761904762
